### DISCUSSION TITLE
MX vs numpy inputs - Performance difference

### DISCUSSION THREAD

**Discussion by sandeep026** [casadi/casadi#3388] [Q&A] [answered]
I am seeing considerable difference when MX and numpy vectors are used as inputs in the map function.

I am using the parallel.py  given in the example pack for the test. The dummy input is provided as a numpy array by default. However, when I  typecast it to MX array, the code runs several orders of magnitude faster. (I have also increased the number of times for the map to be evaluated from  300 to 5000 ) Why is this happening?

[npvsMX.py.zip](https://github.com/casadi/casadi/files/12738557/npvsMX.py.zip)

**Comment by brunoml5** [comment]
Hi! I believe that every time you call a casadi function with numpy, it needs to be cast as a DM object, and only then the code runs. If you need more speed, try working only with DM matrices in your script and only convert to numpy array at the end. It might help.

PS: I'm not a developer.

**Reply by sandeep026** [reply]
Ok. I will run more tests and let you know.

**Comment by jgillis** [comment, marked as answer]
This observation is highly misleading. Wrapping the numeric arguments in an MX changes the result from numerical to symbolic: no actual computations are performed, you are merely creating a small computational graph with a symbolic call.
If you evaluate the symbolic results `evalf(fNaiveParallel(MX(dummyInput)))` you are back to the numpy array performance.

**Reply by sandeep026** [reply]
Thanks for the clarification. The time taken for evaluation with the numpy array and the evalf(fNaiveParallel(MX(dummyInput))) are indeed very close.
